### PR TITLE
add StableDiffusionXLControlNetImg2ImgPipeline

### PR DESCRIPTION
**Pull request by yiyixuxu**
working super well now! 

```python
import torch
import matplotlib
import matplotlib.cm
import numpy as np

from PIL import Image

from diffusers.utils import load_image
from diffusers import StableDiffusionXLControlNetImg2ImgPipeline, ControlNetModel, AutoencoderKL



torch.hub.help("intel-isl/MiDaS", "DPT_BEiT_L_384", force_reload=True)  # Triggers fresh download of MiDaS repo
model_zoe_n = torch.hub.load("isl-org/ZoeDepth", "ZoeD_NK", pretrained=True).eval()
model_zoe_n = model_zoe_n.to("cuda")


def colorize(value, vmin=None, vmax=None, cmap='gray_r', invalid_val=-99, invalid_mask=None, background_color=(128, 128, 128, 255), gamma_corrected=False, value_transform=None):
    if isinstance(value, torch.Tensor):
        value = value.detach().cpu().numpy()

    value = value.squeeze()
    if invalid_mask is None:
        invalid_mask = value == invalid_val
    mask = np.logical_not(invalid_mask)

    # normalize
    vmin = np.percentile(value[mask],2) if vmin is None else vmin
    vmax = np.percentile(value[mask],85) if vmax is None else vmax
    if vmin != vmax:
        value = (value - vmin) / (vmax - vmin)  # vmin..vmax
    else:
        # Avoid 0-division
        value = value * 0.

    # squeeze last dim if it exists
    # grey out the invalid values

    value[invalid_mask] = np.nan
    cmapper = matplotlib.cm.get_cmap(cmap)
    if value_transform:
        value = value_transform(value)
        # value = value / value.max()
    value = cmapper(value, bytes=True)  # (nxmx4)

    # img = value[:, :, :]
    img = value[...]
    img[invalid_mask] = background_color

    # gamma correction
    img = img / 255
    img = np.power(img, 2.2)
    img = img * 255
    img = img.astype(np.uint8)
    img = Image.fromarray(img)
    return img


def get_zoe_depth_map(image):
    with torch.autocast("cuda", enabled=True):
        depth = model_zoe_n.infer_pil(image)
    depth = colorize(depth, cmap="gray_r")
    return depth



controlnet = ControlNetModel.from_pretrained(
    "diffusers/controlnet-zoe-depth-sdxl-1.0",
    use_safetensors=True,
    torch_dtype=torch.float16,
).to("cuda")
vae = AutoencoderKL.from_pretrained("madebyollin/sdxl-vae-fp16-fix", torch_dtype=torch.float16).to("cuda")
pipe = StableDiffusionXLControlNetImg2ImgPipeline.from_pretrained(
    "stabilityai/stable-diffusion-xl-base-1.0",
    controlnet=controlnet,
    vae=vae,
    variant="fp16",
    use_safetensors=True,
    torch_dtype=torch.float16,
).to("cuda")
pipe.enable_model_cpu_offload()


prompt = "A robot, 4k photo"
negative_prompt = "lowres, text, error, cropped, worst quality, low quality, jpeg artifacts, ugly, duplicate, morbid, mutilated, out of frame, extra fingers, mutated hands, poorly drawn hands, poorly drawn face, mutation, deformed, blurry, dehydrated, bad anatomy, bad proportions, extra limbs, cloned face, disfigured, gross proportions, malformed limbs, missing arms, missing legs, extra arms, extra legs, fused fingers, too many fingers, long neck, username, watermark, signature"

image = load_image(
             "https://huggingface.co/datasets/hf-internal-testing/diffusers-images/resolve/main"
            "/kandinsky/cat.png"
        ).resize((1024, 1024))

controlnet_conditioning_scale = 0.55

depth_image = get_zoe_depth_map(image).resize((1024, 1024))

generator = torch.Generator("cuda").manual_seed(0)
images = pipe(
    prompt, image = image, control_image=depth_image, strength=0.99, num_inference_steps=50, controlnet_conditioning_scale=controlnet_conditioning_scale, generator=generator
).images

images[0].save(f"sdxl_robot_cat.png")
```
robot_cat_sdxl_img2img

as a comparison, this is the output we would get from regular controlnet 😂😂😂

robot_cat_not_img2img

### CODE REVIEW

**Comment by yamkz** [comment]
😍
https://github.com/huggingface/diffusers/issues/4589

**Comment by bghira** [comment]
please try:

```py
seed = 123456
generator_1 = torch.manual_seed(seed)
generator_2 = torch.manual_seed(seed ^ 0xFFFFFFF)
```

or similar, to avoid using the same seed for both pipelines. this might improve the quality.

the other thing is that Img2Img guidance scale needs to be lower. it defaults to 5.0. maybe this is too low for use with ControlNet. 

when you say it doesn't work well, you mean in any context using base SDXL 1.0 as img2img model?

**Comment by bghira** [comment]
"doesn't work" means it kicks out an error, or simply results in unexpected image?

**Comment by bghira** [comment]
i think in your example you just provided the only issue i can see is the input image resolution is too low.

**Comment by yiyixuxu** [comment]
> i think in your example you just provided the only issue i can see is the input image resolution is too low.

so pass a `original_size` or`target_size` should help?

**Comment by bghira** [comment]
i don't believe that will help, it just seems that SDXL's ability to generate low resolution images is compromised due to its fine-tuning on 1024px. it is as if the transformer layers are heavily conditioned to represent details that simply won't exist in lower resolution.

**Comment by yiyixuxu** [comment]
@bghira 
> i don't believe that will help, it just seems that SDXL's ability to generate low resolution images is compromised due to its fine-tuning on 1024px. it is as if the transformer layers are heavily conditioned to represent details that simply won't exist in lower resolution.

ok, rechecked the paper, they did mention this. will try it on a 1024 x 1024 image

**Comment by yiyixuxu** [comment]
@bghira 

So I used this one image generated from SDXL text2imge. It's kind of worked a little bit but way worse than the SD1.5 output

original image(generated from SDXL) 
yiyi_test_1_init_image
SDXL img2img
yiyi_test_1_out_fantasy_landscape_sdxl
SD1.5 img2img
yiyi_test_1_out_fantasy_landscape_sd15

**Comment by bghira** [comment]
Sorry for my chosen example image, but I wanted to see if I could reproduce the issue, or if it would behave like 'hires fix':

The input, a classic meme:
image

The output:
image

I looped it back through img2img about 60+ times:
image

My scheduler configuration:

```json
{
  "_class_name": "DDIMScheduler",
  "_diffusers_version": "0.19.0.dev0",
  "beta_end": 0.012,
  "beta_schedule": "scaled_linear",
  "beta_start": 0.00085,
  "clip_sample": false,
  "interpolation_type": "linear",
  "num_train_timesteps": 1000,
  "prediction_type": "epsilon",
  "sample_max_value": 1.0,
  "set_alpha_to_one": false,
  "skip_prk_steps": true,
  "steps_offset": 1,
  "timestep_spacing": "trailing",
  "trained_betas": null,
  "use_karras_sigmas": false
  }
```

notably, we used `timestep_spacing='trailing'` via DDIMScheduler at `strength=0.5` and `guidance_scale=7.5`

you can see that the detail is still obliterated from the background of the image, but fine details are still present.

**Comment by yiyixuxu** [comment]
related :https://github.com/huggingface/diffusers/issues/4724

**Review by patrickvonplaten — APPROVED** [review]
Looks great! Can we add some tests here as well? Then it should be good to go :-)

File: src/diffusers/pipelines/controlnet/pipeline_controlnet_sd_xl_img2img.py
Comment (sayakpaul) [inline src/diffusers/pipelines/controlnet/pipeline_controlnet_sd_xl_img2img.py:100]: The example needs to be edited.

File: src/diffusers/pipelines/controlnet/pipeline_controlnet_sd_xl_img2img.py
Comment (sayakpaul) [inline src/diffusers/pipelines/controlnet/pipeline_controlnet_sd_xl_img2img.py:155]: Where is this coming from?

File: src/diffusers/pipelines/controlnet/pipeline_controlnet_sd_xl_img2img.py
Comment (sayakpaul) [inline src/diffusers/pipelines/controlnet/pipeline_controlnet_sd_xl_img2img.py:162]: We support Multi ControlNet for SDXL. https://huggingface.co/docs/diffusers/main/en/api/pipelines/controlnet_sdxl#multicontrolnet

File: src/diffusers/pipelines/controlnet/pipeline_controlnet_sd_xl_img2img.py
Comment (sayakpaul) [inline src/diffusers/pipelines/controlnet/pipeline_controlnet_sd_xl_img2img.py:577]: Very rigorous.

File: src/diffusers/pipelines/controlnet/pipeline_controlnet_sd_xl_img2img.py
Comment (sayakpaul) [inline src/diffusers/pipelines/controlnet/pipeline_controlnet_sd_xl_img2img.py:725]: When did we add support aesthetic scoring?

File: src/diffusers/pipelines/controlnet/pipeline_controlnet_sd_xl_img2img.py
Comment (sayakpaul) [inline src/diffusers/pipelines/controlnet/pipeline_controlnet_sd_xl_img2img.py:834]: Have you tested their effect?

**Review by sayakpaul — COMMENTED** [review]
Thanks for adding this. 

* Let's fix the example usage and add an entry to the docs. 
* Tests. 
* Could you shed more light on aesthetic scores? I don't think we support it yet nor have we run any experiments to validate its effectivity.

File: src/diffusers/pipelines/controlnet/pipeline_controlnet_sd_xl_img2img.py
Comment (bghira) [reply src/diffusers/pipelines/controlnet/pipeline_controlnet_sd_xl_img2img.py:725]: this is copied from the XL Img2Img pipeline, which is designed for use with the refiner.

i suppose you could tune a ControlNet with aesthetic values, and it would:

* lead to lower prompt adherence, in favour of aesthetic scored data distribution
* require conditional dropout so that we don't overfit to relying on these scores
* need additional dataset management

that said, i would actually like to see support for refiner tuning.

Comment (yiyixuxu) [reply src/diffusers/pipelines/controlnet/pipeline_controlnet_sd_xl_img2img.py:809]: I think aesthetic scores won't actually take effect here since the `config.requires_aesthetics_score` is set to `False`. I'm slightly in favor of keeping these arguments because:
1. one can potentially train a controlnet with refiner in the future (maybe?) 
2. It is consistent with sdxl img2img pipeline, where aesthetic scores are also just place holders when you use it with base model. 

however, I don't have a strong opinion on this, will be happy to remove it if you guys think it's better that way

Comment (patrickvonplaten) [reply src/diffusers/pipelines/controlnet/pipeline_controlnet_sd_xl_img2img.py:809]: Agree with your judgement here @yiyixuxu - I also think it could make sense to train a controlnet refiner (by limiting the added noise to < 20%)

File: src/diffusers/pipelines/controlnet/pipeline_controlnet_sd_xl_img2img.py
Comment (patrickvonplaten) [reply src/diffusers/pipelines/controlnet/pipeline_controlnet_sd_xl_img2img.py:184]: BTW let's make sure to add a docstring for `requires_aestetcs_score` here as well

File: src/diffusers/pipelines/controlnet/pipeline_controlnet_sd_xl_img2img.py
Comment (patrickvonplaten) [inline src/diffusers/pipelines/controlnet/pipeline_controlnet_sd_xl_img2img.py:1287]: Can we update this part of the code to conform with the changes here: https://github.com/huggingface/diffusers/pull/4796 ?

**Comment by patrickvonplaten** [comment]
Cool I think we're mostly good to go here no? Ok to merge for me once all tests are green. @sayakpaul if you could give a final review that'd be great!

File: src/diffusers/pipelines/controlnet/pipeline_controlnet_sd_xl_img2img.py
Comment (sayakpaul) [reply src/diffusers/pipelines/controlnet/pipeline_controlnet_sd_xl_img2img.py:184]: I think this is yet to be addressed.

File: src/diffusers/pipelines/controlnet/pipeline_controlnet_sd_xl_img2img.py
Comment (sayakpaul) [inline src/diffusers/pipelines/controlnet/pipeline_controlnet_sd_xl_img2img.py:1229]: I guess we need to also correct the behaviour for `guess_mode` here per your other PR?

File: tests/pipelines/controlnet/test_controlnet_sdxl_img2img.py
Comment (sayakpaul) [inline tests/pipelines/controlnet/test_controlnet_sdxl_img2img.py:199]: What is the problem here?

**Review by sayakpaul — APPROVED** [review]
Looking very nice! Thanks for this one!

**Comment by sayakpaul** [comment]
Let's also fix the tests before merging :-)

File: tests/pipelines/controlnet/test_controlnet_sdxl_img2img.py
Comment (yiyixuxu) [reply tests/pipelines/controlnet/test_controlnet_sdxl_img2img.py:199]: not sure actually! I thought maybe you left the note 😂
it was copied from sdxl-img2img

**Comment by markrmiller** [comment]
Ignore if I'm offbase for some reason, but I'm not sure from_single_file support is there.

File: tests/pipelines/controlnet/test_controlnet_sdxl_img2img.py
Comment (patrickvonplaten) [reply tests/pipelines/controlnet/test_controlnet_sdxl_img2img.py:223]: Ah yeah that note was me! The problem is that for img2img/refiner the `"text_encoder"` is optional so we should be able to load the pipeline without it, but that's not possible when using the non-refiner architecture (which is currently done in tests). We should probably add a new test class here at some point

File: src/diffusers/pipelines/controlnet/pipeline_controlnet_sd_xl_img2img.py
Comment (richardSHkim) [inline src/diffusers/pipelines/controlnet/pipeline_controlnet_sd_xl_img2img.py:1201]: @yiyixuxu Thank you for your great work, but I think `height, width = control_image.shape[-2:]` should be `height, width = control_image[0].shape[-2:]` when controlnet is `MultiControlNetModel` as in  [StableDiffusionXLControlNetPipeline](https://github.com/huggingface/diffusers/blob/8ccb619416a36f5951dac654a92e869d76db4bbc/src/diffusers/pipelines/controlnet/pipeline_controlnet_sd_xl.py#L1035).

File: tests/pipelines/controlnet/test_controlnet_sdxl_img2img.py
Comment (polavishnu4444) [reply tests/pipelines/controlnet/test_controlnet_sdxl_img2img.py:223]: When is the plan for the refiner support of controlnet Img2Img planned for?

**Pull request merged**
